### PR TITLE
Add welcome wizard slides on first start requesting permissions

### PR DESCRIPTION
**Pull request by ghost**
Purpose
Improve the app by adding a welcome wizard explaining required and optional permissions.

Screenshots
image
image
image
image

Testing
Verified working on device lg-h815 running Android 7.1.2 at commit https://github.com/syncthing/syncthing-android/pull/1176/commits/fd490d8a1f87cc08bddfc6129c857b0047ff569a .

EDIT: Removed the relation to issue #1129 . I'll do that fix based on this PR in a follow-up so this doesn't get too big.

### CODE REVIEW

**Comment by AudriusButkevicius** [comment]
Its all in german so I have no idea what the guide says.

**Comment by Catfriend1** [comment]
@AudriusButkevicius see strings.xml for the English texts. The german parts got loaded because I reused the texts that syncthing already had for german since many releases.

**Comment by Catfriend1** [comment]
@AudriusButkevicius The code for the slides requesting permissions from the user is not huge, so I decided to add it to FirstStartActivity (which was nearly "empty" before). It's all in one place now :).

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java
Comment (Catfriend1) [inline app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: Sidenote: I've replaced the "isFirstStart" check (previously done via preference check) because it's not very reliable detecting if syncthing needs to generate new keys and config. We now use the same indicator we already have in other classes (e.g. ConfigXml). If no config is present (or it got corrupted in a way crucial files are missing) we'll trigger the "key and config generation" process on app start. This gracefully auto-fixes an unrecoverable config break so we don't have to advise users to clear app's data and cache.

File: app/src/main/res/values/strings.xml
Comment (imsodin) [inline app/src/main/res/values/strings.xml]: As this is a required permission, I wouldn't ask and have a button - just explain and then request the permission on pressing "next".

Alternative wording that doesn't just repeat the title and seems "clearer" and easier to me:  
`Syncthing needs to access your storage to do file synchronization.`

File: app/src/main/res/values/strings.xml
Comment (imsodin) [inline app/src/main/res/values/strings.xml]: Potentially capitalize fully, as it's a title.

File: app/src/main/res/values/strings.xml
Comment (imsodin) [inline app/src/main/res/values/strings.xml]: Also maybe capitalize.

File: app/src/main/res/values/strings.xml
Comment (imsodin) [inline app/src/main/res/values/strings.xml]: Here the button makes sense, as it's voluntary. I'd turn the latter part around to focus on that choice instead of the android requirement. Something like `If you want to use this feature, press the button above to give the required location access to Syncthing. Otherwise you can skip this step.`

**Review by imsodin — COMMENTED** [review]
If you have high-level stuff that you think would benefit from another set of eyes, you can ping me. I am not the ideal candidate for it, as I am not very invested in advance functionalities of the app (I am happy as long as it starts and stops when I manually tell it to). I don't follow the syncthing-android activities closely and have no plans to dig into it, but as far as I have time and something to say I am happy to provide feedback. Just to have said it explicitly: Consider all of my feedback as suggestions - I don't have any stakes in the android wrapper.

I believe this is a good improvement, it's always nice to be up front about required permissions. Are these pages conditional, i.e. only show up if the permissions weren't already granted?

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java
Comment (AudriusButkevicius) [inline app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: What are these magic numbers?

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java
Comment (AudriusButkevicius) [inline app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: Are we going to show all slides if the permission is revoked?

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java
Comment (AudriusButkevicius) [inline app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: Why no m prefix for these?

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java
Comment (AudriusButkevicius) [inline app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: What does finish do in this case?

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java
Comment (AudriusButkevicius) [inline app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: capitalisation

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java
Comment (AudriusButkevicius) [inline app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: If it's null, what should we do in that case?

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java
Comment (AudriusButkevicius) [inline app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: Why do we need to override this?

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java
Comment (AudriusButkevicius) [inline app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: Same "why"

File: app/src/main/res/values/colors.xml
Comment (AudriusButkevicius) [inline app/src/main/res/values/colors.xml]: Is there harm in having these uncommented if there are only 3 layouts?

File: app/src/main/res/values/strings.xml
Comment (AudriusButkevicius) [reply app/src/main/res/values/strings.xml]: I think this summons a separate system dialog, hence has to be "initiated" somehow.

Agree with the better wording.

File: app/src/main/res/values/strings.xml
Comment (AudriusButkevicius) [reply app/src/main/res/values/strings.xml]: I think this needs an explanation why...


> Syncthing can be configured to run only on selected wifi networks. Sadly, Android requires applications to have location permissions to be able look up active WiFi network name, as you can sometimes infer users location from the name of the network they are connected to. If you want to use this feature, press the button above to give the required location permissions to Syncthing. Otherwise you can skip this step.

Comment (Catfriend1) [reply app/src/main/res/values/strings.xml]: Hmm.. both suggestions are good. imsodin's is comforting the user but the user doesn't really get why a location permission should be granted for a wifi feature. Audrius' suggestion tells the truth (what for example complies with fdroid policy for app approval) but also looks a tiny bit scary as we say what nasty things we could do (but won't do :). For the first try on "later" public feedback I'll take Audrius' suggestion.

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java
Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: This is an exception in the wizard as we have to make sure the app quits if the user didn't grant storage permissions during the second slide and tell him when exiting why we do it (that was already in the original welcome screen). So it is a slide index to make it easier if we need to rearrange slides later. I chose this way as the ViewPager doesn't really help here returning something unique and easily available to get the xml_layout_slide_name.

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java
Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: Yes, I decided to do so in the first proposal as I consider it rare that users revoke the permission if they actively use the app. (see rare one bug report on the tracker where one user did). Reason is it will never harm anything going through the slides again - also being able to grant the location permission if this was also revoked "later" after completing the wizard for the first time. Cluttering the code to handle rare cases is a maybe can do but currently don't see necessary.

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java
Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: I've had them temporarily in another context, added the "m" and gave the buttons a new var name.

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java
Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: I exits the app by ending the FirstStartActivity like it was the case before the PR if the user did not grant the required storage permission.

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java
Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: ok

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java
Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: We do nothing as this is intended behaviour. The button exists only on one slide (slide 2) so when for example slide 3 gets instantiated it WILL return null as it doesn't exist. In this case we don't register the onClickListener as we don't need it.

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java
Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: The method  instantiateItem(ViewGroup, int) returns an Object for a particular view. PagerAdapter implementation is considering this Object as a key value when viewpager changes a page. If we return the view itself from instantiateItem(ViewGroup, int), then our key for that page becomes the view itself. We can check return view == obj; from isViewFromObject which will always return true and our pages will display.

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java
Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: Proper uninitialization and the above mentioned reason.

File: app/src/main/res/values/colors.xml
Comment (Catfriend1) [reply app/src/main/res/values/colors.xml]: Uncommenting will throw a lint warning but not be a problem. I'll reserved the color as I already know what the fourth slide will be.

**Comment by AudriusButkevicius** [comment]
Other than the next button exiting if storage permission is not granted, looks good to me.

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java
Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: Android development guidelines for permission request say we're not allowed to ask this multiple times on our own. That's why the button makes sense.

**Comment by Catfriend1** [comment]
@AudriusButkevicius Okay, then let's go. As we currently have many discussion I'd suggest a follow-up PR would be great to meet again making the "app exits when storage permission not granted" better.

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java
Comment (AudriusButkevicius) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: We should just disable the next button if the user has not granted permissions.

Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: A message is already printed on the screen when exiting, so it's no crash.
Disabling the next button would leave no way to exit the foreground app window. User would then have to press home or the panic button.

Comment (AudriusButkevicius) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: Why can't he press the back button?

Comment (AudriusButkevicius) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: Or we could just make clicking next print the toast and not progress to the new window. Then the user can back out, or grant the permission.

Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: I'll think about that. Good for follow-up.

**Review by AudriusButkevicius — DISMISSED** [review]

**Comment by AudriusButkevicius** [comment]
Anything left todo, or are we good to merge this?

**Comment by Catfriend1** [comment]
Ok from my side. (I'll already mentioned I'd like to take care of more welcome slides, especially those required to make the user experience smooter on first app start relating to key generation and run conditions).

@imsodin: Is is okay from your side? Some of your feedback already got in.

File: app/src/main/res/values/strings.xml
Comment (imsodin) [reply app/src/main/res/values/strings.xml]: Fine with me, but please drop the sadly. We want the user to be cheerful about using Syncthing ;)

**Review by imsodin — CHANGES_REQUESTED** [review]
I left my notes on an outdated review, meaning they are hidden. Thus to repeat here:

The location permission explanation is fine with me, but please drop the sadly. We want the user to be cheerful about using Syncthing ;)

And I didn't mean to put the strings in all caps, but to capitalize them, i.e. make the first letter uppercase: Storage permissions -> Storage Permissions (again, that's just how I am used to do things for title-like properties, maybe mixed capitalization looks better. I don't care much about it, except that it is **not** all caps :)

**Comment by Catfriend1** [comment]
Thanks for reviewing, @imsodin I got all comments via mail that were hidden and will take care later.

**Comment by Catfriend1** [comment]
@AudriusButkevicius Would you mind leaving the merges up to me after a PR has got the required approvements? Then I would plan to do a final testing after the reviews and before hitting the merge.

**Comment by AudriusButkevicius** [comment]
Sure, but everytime you merge something in the branch the approval gets dismissed.

**Review by AudriusButkevicius — APPROVED** [review]

**Review by imsodin — APPROVED** [review]

**Comment by Catfriend1** [comment]
Ok, testing completed, no crashes or problems so far found.

**Pull request merged**
